### PR TITLE
Fix DeadLock in FdManager::ChangeEntityToTempPath

### LINKED ISSUE #2454: dead lock in NoCacheLoadAndPost
#### Version of s3fs being used (`s3fs --version`)

1.94

```

### Details about issue
<!-- Please describe the content of the issue in detail. -->
Thread 8 (Thread 0x7f41a1e4e700 (LWP 23788)):
#0  0x00007f41a664854d in __lll_lock_wait () from /lib64/libpthread.so.0
#1  0x00007f41a6643e9b in _L_lock_883 () from /lib64/libpthread.so.0
#2  0x00007f41a6643d68 in pthread_mutex_lock () from /lib64/libpthread.so.0
#3  0x000055bf4d169b39 in AutoLock::AutoLock(pthread_mutex_t*, AutoLock::Type) ()
#4  0x000055bf4d14fe14 in FdEntity::Clear() ()
#5  0x000055bf4d1502de in FdEntity::~FdEntity() ()
#6  0x000055bf4d149588 in FdManager::ChangeEntityToTempPath(FdEntity*, char const*) ()
#7  0x000055bf4d150c95 in FdEntity::NoCacheLoadAndPost(PseudoFdInfo*, long, long) ()
#8  0x000055bf4d1518a8 in FdEntity::WriteMixMultipart(PseudoFdInfo*, char const*, long, unsigned long) ()
#9  0x000055bf4d15202d in FdEntity::Write(int, char const*, long, unsigned long) ()
#10 0x000055bf4d0e6f82 in s3fs_write(char const*, char const*, unsigned long, long, fuse_file_info*) ()
#11 0x00007f41a7cbea7c in fuse_fs_write_buf () from /lib64/libfuse.so.2
#12 0x00007f41a7cbebf8 in fuse_lib_write_buf () from /lib64/libfuse.so.2
#13 0x00007f41a7cc7cf8 in fuse_ll_process_buf () from /lib64/libfuse.so.2
#14 0x00007f41a7cc4401 in fuse_do_work () from /lib64/libfuse.so.2
#15 0x00007f41a6641ea5 in start_thread () from /lib64/libpthread.so.0
#16 0x00007f41a636a96d in clone () from /lib64/libc.so.6
```
lock is already taken in  ` FdEntity::Write`
image
image

### PR DESCRIPTION
**Pull request by VVoidV**
this fix #2454

### CODE REVIEW

File: src/fdcache.cpp [hunk 1]
```diff
@@ -783,8 +783,10 @@ bool FdManager::ChangeEntityToTempPath(FdEntity* ent, const char* path)
         if(iter->second.get() == ent){
             std::string tmppath;
             FdManager::MakeRandomTempPath(path, tmppath);
-            iter->second.reset(ent);
-            break;
+            // Move the entry to the new key
+            fent[tmppath] = std::move(iter->second);
+            iter = fent.erase(iter);
+            return true;
```
Comment (ggtakec) [inline src/fdcache.cpp:789]: Thanks for this line.

**Review by ggtakec — CHANGES_REQUESTED** [review]
Thank you for finding this bug!
I think there is no problem with your modification method and content.

However, since an error occurred during the cppcheck inspection, it is necessary to change the prototype of the method below.

- src/fdcache.cpp  
```
-bool FdManager::ChangeEntityToTempPath(FdEntity* ent, const char* path)
+bool FdManager::ChangeEntityToTempPath(const FdEntity* ent, const char* path)
```
- src/fdcache.h  
```
- bool ChangeEntityToTempPath(FdEntity* ent, const char* path);
+ bool ChangeEntityToTempPath(const FdEntity* ent, const char* path);
```

Could you please update(and rebase) this fix?

**Comment by VVoidV** [comment]
@ggtakec  update and rebased

**Review by ggtakec — APPROVED** [review]

**Comment by ggtakec** [comment]
@VVoidV Thank you for quick response. I've merged this.

**Pull request merged**